File: plot_time_series.py
'''
Description: 
Author: Ming Fang
Date: 2024-05-20 15:12:02
LastEditors: Ming Fang
LastEditTime: 2024-05-20 15:27:27
'''
import h5py
from matplotlib import pyplot as plt
import numpy as np


if __name__ == "__main__":
    # Load the data
    with h5py.File("test/coincidences.h5", "r") as f:
        coincident_pulses = f["voltage_pulses"][:]

    for coincidence_index in range(20):
        # Plot the data
        fig, ax = plt.subplots()
        for i in range(5):
            time_series = coincident_pulses[coincidence_index,i, :]
            ax.plot(2*np.arange(len(time_series)), time_series, label=f"Channel {i}")
        ax.set_xlabel("Time (ns)")
        ax.set_ylabel("Voltage (V)")
        ax.legend()
        plt.show()

File: plot_time_series.m
% Main script
coincidence_index = 1;

% Load the data
filename = 'test/coincidences.h5';
coincident_pulses = h5read(filename, '/voltage_pulses');
coincident_pulses = permute(coincident_pulses, [3 2 1]);
time_stamps = h5read(filename, '/time_stamps');
time_stamps = transpose(time_stamps);

% % Plot the data
% figure;
% hold on;
% time_series = create_time_series(squeeze(coincident_pulses(coincidence_index, :, :)), time_stamps(coincidence_index, :));
% for i = 1:5
%     plot(2*(0:(size(time_series, 2)-1)), time_series(i, :), 'DisplayName', sprintf('Channel %d', i-1));
% end
% xlabel('Time (ns)');
% ylabel('Voltage (V)');
% legend;
% hold off;

% Loop through the first 20 coincidences and plot the data
for coincidence_index = 1:20
    % Create the time series
    time_series = create_time_series(squeeze(coincident_pulses(coincidence_index, :, :)), time_stamps(coincidence_index, :));
    
    % Plot the data
    figure;
    hold on;
    for i = 1:5
        plot(2*(0:(size(time_series, 2)-1)), time_series(i, :), 'DisplayName', sprintf('Channel %d', i-1));
    end
    xlabel('Time (ns)');
    ylabel('Voltage (V)');
    legend;
    hold off;
end

function time_series = create_time_series(pulses, time_stamps)
    % start_time = min(time_stamps);
    % disp(strcat('First trigger at: ', string(start_time), ' ps'));
    % disp(strcat('Triggering time of 5 channels with repspect to the first trigger: ', num2str((time_stamps - start_time)), ' ps'));
    % shift_samples = floor((time_stamps - start_time) / 2000); % 2000 ps is the sampling step
    % time_series_length = max(shift_samples) + size(pulses, 2);

    % time_series = zeros(size(pulses, 1), time_series_length);
    % for i = 1:size(pulses, 1)
    %     time_series(i, 1:shift_samples(i)) = pulses(i, 1);
    %     time_series(i, shift_samples(i)+1:shift_samples(i)+size(pulses, 2)) = pulses(i, :);
    %     time_series(i, shift_samples(i)+size(pulses, 2)+1:end) = pulses(i, end);
    % end
    time_series = pulses;
end
